From 70156fd28b6d3fbb365e28106e7bf4ddcc80c62d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Dave=20T=C3=A4ht?= <dave.taht@bufferbloat.net>
Date: Sun, 9 Jun 2013 11:19:40 -0700
Subject: [PATCH] Handle encapsulated ipv6 better

The flow dissector did not handle ipv6 extended headers properly
---
 net/core/flow_dissector.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/net/core/flow_dissector.c b/net/core/flow_dissector.c
index 25ebce7..a14bd91 100644
--- a/net/core/flow_dissector.c
+++ b/net/core/flow_dissector.c
@@ -54,6 +54,8 @@ ip:
 	case __constant_htons(ETH_P_IPV6): {
 		const struct ipv6hdr *iph;
 		struct ipv6hdr _iph;
+		__be16 frag = 0;
+		int nhoff6; 
 ipv6:
 		iph = skb_header_pointer(skb, nhoff, sizeof(_iph), &_iph);
 		if (!iph)
@@ -63,6 +65,9 @@ ipv6:
 		flow->src = (__force __be32)ipv6_addr_hash(&iph->saddr);
 		flow->dst = (__force __be32)ipv6_addr_hash(&iph->daddr);
 		nhoff += sizeof(struct ipv6hdr);
+		if((nhoff6 = ipv6_skip_exthdr(skb,nhoff,&ip_proto,&frag)) > -1)
+			nhoff += nhoff6;
+		if(frag != 0) ip_proto = 0;
 		break;
 	}
 	case __constant_htons(ETH_P_8021Q): {
-- 
1.8.1.2

